ciao-controller/cnci.go

```diff
@@ -0,0 +1,579 @@
+type CNCIManager struct {
+	tenant string
+	ctrl   *controller
+
+	// there's no reason to have separate lock for each map.
+	cnciLock *sync.RWMutex
+
+	// this is a map of CNCI instance IDs to CNCI structs
+	cncis map[string]*CNCI
+
+	// this is a map of subnet (integer) to CNCI structs
+	subnets map[int]*CNCI
+}
+
+func (c *CNCI) stop() error {
+	err := transitionInstanceState(c.instance, payloads.Stopping)
+	if err != nil {
+		return err
+	}
+
+	err = c.ctrl.deleteInstance(c.instance.ID)
+	if err != nil {
+		return errors.Wrapf(err, "error deleting CNCI instance")
+	}
+
+	return nil
+}
+
+func waitForEventTimeout(ch chan event, e event, timeout time.Duration) error {
+	select {
+	case recv := <-ch:
+		if recv != e {
+			return fmt.Errorf("expecting %v got %v", e, recv)
+		}
+		return nil
+	case <-time.After(timeout):
+		return fmt.Errorf("timeout waiting for event %v", e)
+	}
+}
+
+func (c *CNCI) transitionState(to CNCIState) {
+	transitionInstanceState(c.instance, (string(to)))
+
+	// some state changes cause events
+	c.eventChLock.Lock()
+	ch := c.eventCh
+	c.eventChLock.Unlock()
+
+	if ch == nil {
+		return
+	}
+
+	switch to {
+	case exited:
+		*ch <- removed
+	case active:
+		*ch <- added
+	case failed:
+		*ch <- startFailure
+	}
+}
+
+// Active will return true if the CNCI has been launched successfully
+func (c *CNCIManager) Active(ID string) bool {
+	c.cnciLock.RLock()
+	defer c.cnciLock.RUnlock()
+
+	cnci, ok := c.cncis[ID]
+	if !ok {
+		return false
+	}
+
+	return instanceActive(cnci.instance)
+}
+
+func (c *CNCIManager) launch(subnet string) (*types.Instance, error) {
+	glog.V(2).Infof("launching cnci for subnet %s", subnet)
+
+	workloadID, err := c.ctrl.ds.GetCNCIWorkloadID()
+	if err != nil {
+		return nil, err
+	}
+
+	w := types.WorkloadRequest{
+		WorkloadID: workloadID,
+		TenantID:   c.tenant,
+		Instances:  1,
+		Subnet:     subnet,
+		Name:       fmt.Sprintf("CNCI for subnet %s", subnet),
+	}
+
+	instances, err := c.ctrl.startWorkload(w)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to Launch CNCI")
+	}
+
+	return instances[0], nil
+}
+
+// WaitForActive will launch a cnci if needed and wait for it to be active,
+// or wait for an existing cnci to become active.
+func (c *CNCIManager) WaitForActive(subnet int) error {
+	c.cnciLock.Lock()
+
+	cnci, ok := c.subnets[subnet]
+	if ok {
+		if cnci.timer != nil {
+			cnci.timer.Stop()
+			cnci.timer = nil
+		}
+
+		// subnet already exists
+		c.cnciLock.Unlock()
+
+		// block until subnet is active
+		return c.waitForActive(subnet)
+	}
+
+	glog.V(2).Infof("cnci does not exist for subnet %s", subnet)
+
+	ch := make(chan event)
+
+	cnci = &CNCI{
+		ctrl:        c.ctrl,
+		eventCh:     &ch,
+		eventChLock: sync.Mutex{},
+		subnet:      subnet,
+	}
+
+	// we initialized the eventCh because we are going to wait for
+	// an event. Close and delete at the conclusion of this function.
+	defer func() {
+		cnci.eventChLock.Lock()
+		close(ch)
```

Comment:
No need to close the channel as far as I can tell.  By the time we get here there will be nobody reading from it.

Reply:
Okay so maybe I'm wrong.  If we have two parallel calls to waitForTimeout we need the close.